### ISSUE TITLE
Should `scale` and `bias` be required inputs for `batchNormalization` op?

### ISSUE THREAD

**Issue by huningxin** [webmachinelearning/webnn#481]
(Thanks @wacky6 for raising this issue during reviewing Chromium [CL-5034594](https://chromium-review.googlesource.com/c/chromium/src/+/5034594/comment/8a90a4b5_b3e34f1e/))

Regarding to the existing [`batchNormalization`](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-batchnorm) definition, `scale` and `bias` operands are optional members of [`MLBatchNormalizationOptions`](https://www.w3.org/TR/webnn/#dictdef-mlbatchnormalizationoptions) dictionary.  However, `scale` (a.k.a `gamma`) and `bias` (a.k.a `beta`) are usually required inputs across frameworks and native ML APIs.

### Frameworks:
* TensorFlow's [`tf.nn.batch_normalization`](https://www.tensorflow.org/api_docs/python/tf/nn/batch_normalization):  `offset` (equivalent to `bias`) and `scale` are required parameters.
* ONNX's [`BatchNormalization`](https://onnx.ai/onnx/operators/onnx__BatchNormalization.html): `scale` and `B` (equivalent to `bias`) are required inputs.

### Native ML APIs:
* DirectML's [`DML_BATCH_NORMALIZATION_OPERATOR_DESC`](https://learn.microsoft.com/en-us/windows/win32/api/directml/ns-directml-dml_batch_normalization_operator_desc): `ScaleTensor` and `BiasTensor` are **not** annotated with `_Maybenull_ `, so they are supposed to be required. 
* MPS's [`MPSCNNBatchNormalizationDataSource`](https://developer.apple.com/documentation/metalperformanceshaders/mpscnnbatchnormalizationdatasource?language=objc): `beta` (equivalent to `bias`) and `gamma` (equivalent to `scales`) are annotated with `Required`.
* Stable HLO's [`batch_norm_inference`](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#batch_norm_inference): `scale` and `offset` (equivalent to `bias`) are required inputs.

The proposal is to make the two operands required , for example
```
dictionary MLBatchNormalizationOptions {
  unsigned long axis = 1;
  float epsilon = 1e-5;
  MLActivation activation;
};

partial interface MLGraphBuilder {
  MLOperand batchNormalization(MLOperand input, MLOperand mean, MLOperand variance,
                               MLOperand scale, MLOperand bias,
                               optional MLBatchNormalizationOptions options = {});
};
```

For some models that won't use `scale` and `bias` at the inference time, e.g., DenseNet 121, the frameworks can set `scale`'s values to 1 and `bias`'s values to 0.

/cc @wchao1115 @fdwr

**Comment by fdwr** [comment]
@huningxin : Your analysis is persuasive. It might be convenient to allow scale and bias to be optional, but if underlying backends do not support it (forcing implementations to add dummy 0 and 1 tensors), and frameworks are unlikely to generate such a call anyway, then making them required makes sense to me. (and yes, your reading of DML_BATCH_NORMALIZATION_OPERATOR_DESC is correct).

**Comment by wchao1115** [comment]
@huningxin If I read this correctly, are you saying that tensor params should never be optional b/c it causes the implementation to have to allocate unnecessary buffer resources for them when dealing with a platform API that already treats them as required?

**Comment by huningxin** [comment]
@wchao1115 

> @huningxin If I read this correctly, are you saying that tensor params should never be optional b/c it causes the implementation to have to allocate unnecessary buffer resources for them when dealing with a platform API that already treats them as required?

The buffer resources are less concerned, because I suppose frameworks have to allocate dummy 0 and 1 tensors if the models don't need scale and bias, like DenseNet. 

My point is if majority frameworks and native ML APIs have the scale and bias required, WebNN might be worth aligning with that, because this would simplify WebNN's implementation to deal with this uncommon usages.

However, as I mentioned in last WG call, on the other hand, this may prevent the potential future optimizations that a native implementation may eliminate the unnecessary element-wise multiplication (for scale) and addition (for bias) if the two are not present. So, I am wondering whether there is a such plan to do that optimization in native implementations. We may want to make this interface future-proof.

**Comment by fdwr** [comment]
> We may want to make this interface future-proof.

One can always relax a tighter constraint (WebNN requiring scale and bias) to a looser constraint later (them being optional). So I don't see we're precluding any future optimizations by starting with the tighter constraint (making them required).

**Comment by huningxin** [comment]
CoreML [mil.ops.defs.iOS15.normalization.batch_norm](https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS15.normalization.batch_norm) allows beta (equivalent to bias) and gamma (equivalent to scale) to be optional.
